From a63652174985ba4f65a1c249c15e675fdb7a20d7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tim-Philipp=20M=C3=BCller?= <tim@centricular.com>
Date: Sat, 30 May 2020 13:24:46 +0100
Subject: [PATCH 2/2] meson: add cpu-features-path option for Android

Add option to include cpu-features.[ch] from a given path
into the build for platforms that don't provide this out
of the box. This is needed on Android.

We copy the files into the build directory instead of using
them directly from the original location, for better build
hygiene, and so that they are automatically included as part
of CI build artifacts and such.

Need to bump meson dep for fs module.
---
 meson.build        |  2 +-
 meson_options.txt  |  5 +++++
 pixman/meson.build | 23 ++++++++++++++++++++++-
 3 files changed, 28 insertions(+), 2 deletions(-)

diff --git a/meson.build b/meson.build
index a247183..e8e80e8 100644
--- a/meson.build
+++ b/meson.build
@@ -23,7 +23,7 @@ project(
   ['c'],
   version : '0.40.0',
   license : 'MIT',
-  meson_version : '>= 0.50.0',
+  meson_version : '>= 0.53.0',
   default_options : ['buildtype=debugoptimized'],
 )
 
diff --git a/meson_options.txt b/meson_options.txt
index 781e258..355a8f5 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -74,6 +74,11 @@ option(
   type : 'feature',
   description : 'Use GNU style inline assembler',
 )
+option(
+  'cpu-features-path',
+  type : 'string',
+  description : 'Path to platform-specific cpu-features.[ch] for systems that do not provide it (e.g. Android)',
+)
 option(
   'openmp',
   type : 'feature',
diff --git a/pixman/meson.build b/pixman/meson.build
index f48357f..34330d2 100644
--- a/pixman/meson.build
+++ b/pixman/meson.build
@@ -102,6 +102,27 @@ pixman_files = files(
   'pixman-utils.c',
 )
 
+# Android cpu-features
+cpu_features_path = get_option('cpu-features-path')
+cpu_features_sources = []
+if cpu_features_path != ''
+  cpu_features_h_path = join_paths(cpu_features_path, 'cpu-features.h')
+  cpu_features_c_path = join_paths(cpu_features_path, 'cpu-features.c')
+  fs = import('fs')
+  if not fs.exists(cpu_features_c_path) or not fs.exists(cpu_features_h_path)
+    error('Expected cpu-features.[ch] in directory ' + cpu_features_path)
+  endif
+  # copy into build directory
+  cpu_features_h = configure_file(output: 'cpu-features.h',
+    input: cpu_features_h_path,
+    copy: true)
+  cpu_features_c = configure_file(output: 'cpu-features.c',
+    input: cpu_features_c_path,
+    copy: true)
+  message('Using cpu-features.[ch] from ' + cpu_features_path)
+  cpu_features_sources = [cpu_features_c, cpu_features_h]
+endif
+
 # We cannot use 'link_with' or 'link_whole' because meson wont do the right
 # thing for static archives.
 _obs = []
@@ -111,7 +132,7 @@ endforeach
 
 libpixman = library(
   'pixman-1',
-  [pixman_files, config_h, version_h],
+  [pixman_files, config_h, version_h, cpu_features_sources],
   objects : _obs,
   c_args : libpixman_extra_cargs,
   dependencies : [dep_m, dep_threads],
-- 
2.26.1

